File: core/src/main/java/fr/sncf/osrd/external_generated_inputs/ElectricalProfileMapping.java
package fr.sncf.osrd.external_generated_inputs;

import com.google.common.collect.ImmutableRangeMap;
import com.google.common.collect.Range;
import com.google.common.collect.RangeMap;
import com.google.common.collect.TreeRangeMap;
import fr.sncf.osrd.infra_state.api.TrainPath;
import fr.sncf.osrd.railjson.schema.external_generated_inputs.RJSElectricalProfileSet;
import java.util.HashMap;

/**
 * A mapping from track sections to electrical profile values
 * The electrical profiles model the power loss along catenaries depending on the position and the power class of
 * the rolling stock used
 */
public class ElectricalProfileMapping {
    /**
     * Internal representation: {"power class": {"track section": {"range": "electrical profile value"}}}
     */
    protected HashMap<String, HashMap<String, RangeMap<Double, String>>> mapping = new HashMap<>();

    /**
     * Parse the rjs profiles and store them in the internal mapping.
     */
    public void parseRJS(RJSElectricalProfileSet rjsProfileSet) {
        assert mapping.isEmpty();

        for (var rjsProfile : rjsProfileSet.levels) {
            var trackMapping = mapping.computeIfAbsent(rjsProfile.powerClass, k -> new HashMap<>());
            for (var trackRange : rjsProfile.trackRanges) {
                var rangeMapping = trackMapping.computeIfAbsent(trackRange.trackSectionID, k -> TreeRangeMap.create());
                rangeMapping.put(Range.closedOpen(trackRange.begin, trackRange.end), rjsProfile.value);
            }
        }
    }

    /**
     * Returns the electrical profile values encountered along the train path, for each power class given.
     */
    public HashMap<String, RangeMap<Double, String>> getProfilesOnPath(TrainPath trainPath) {
        var res = new HashMap<String, RangeMap<Double, String>>();
        for (var entry : mapping.entrySet()) {
            var powerClass = entry.getKey();
            var byTrackMapping = entry.getValue();
            var rangeMap = new ImmutableRangeMap.Builder<Double, String>();
            double offset = 0;
            for (var trackRange : TrainPath.removeLocation(trainPath.trackRangePath())) {
                var trackID = trackRange.track.getEdge().getID();
                if (byTrackMapping.containsKey(trackID)) {
                    var pathRangeMapping = trackRange.convertMap(byTrackMapping.get(trackID));
                    for (var rangeMapEntry : pathRangeMapping.asMapOfRanges().entrySet())
                        rangeMap.put(Range.closedOpen(rangeMapEntry.getKey().lowerEndpoint() + offset,
                                rangeMapEntry.getKey().upperEndpoint() + offset), rangeMapEntry.getValue());
                }
                offset += trackRange.getLength();
            }
            res.put(powerClass, rangeMap.build());
        }

        return res;
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/kt_external_generated_inputs/ElectricalProfileMapping.kt
package fr.sncf.osrd.kt_external_generated_inputs

import fr.sncf.osrd.railjson.schema.external_generated_inputs.RJSElectricalProfileSet
import fr.sncf.osrd.sim_infra.api.Path
import fr.sncf.osrd.utils.DistanceRangeMap
import fr.sncf.osrd.utils.DistanceRangeMapImpl
import fr.sncf.osrd.utils.units.Distance.Companion.fromMeters

/**
 * A mapping from track sections to electrical profile values
 * The electrical profiles model the power loss along catenaries depending on the position and the power class of
 * the rolling stock used
 */

class ElectricalProfileMapping {
    /**
     * Internal representation: {"power class": {"track section": {"range": "electrical profile value"}}}
     */
    var mapping = HashMap<String, HashMap<String, DistanceRangeMap<String>>>()

    /**
     * Parse the rjs profiles and store them in the internal mapping.
     */
    fun parseRJS(rjsProfileSet: RJSElectricalProfileSet) {
        assert(mapping.isEmpty())
        for (rjsProfile in rjsProfileSet.levels) {
            val trackMapping: HashMap<String, DistanceRangeMap<String>> =
                mapping.computeIfAbsent(
                    rjsProfile.powerClass
                ) { _: String? -> HashMap() }
            for (trackRange in rjsProfile.trackRanges) {
                val rangeMapping = trackMapping.computeIfAbsent(
                    trackRange.trackSectionID
                ) { k: String? -> DistanceRangeMapImpl() }
                rangeMapping.put(fromMeters(trackRange.begin), fromMeters(trackRange.end), rjsProfile.value)
            }
        }
    }

    /** Returns the electrical profiles encountered on a path */
    fun getProfilesOnPath(path: Path): HashMap<String, DistanceRangeMap<String>> {
        val res = HashMap<String, DistanceRangeMap<String>>()
        for (entry in mapping.entries) {
            val powerClass = entry.key
            val byTrackMapping: HashMap<String, DistanceRangeMap<String>> = entry.value
            res.put(powerClass, path.getElectricalProfiles(byTrackMapping))
        }
        return res
    }
}
